### DISCUSSION TITLE
Stochastic Momentum Index (SMI)

### DISCUSSION THREAD

**Discussion by myalgomate** [DaveSkender/Stock.Indicators#625] [Indicators]
Community discussion about [Stochastic Momentum Index (SMI)](https://dotnet.stockindicators.dev/indicators/Smi/#content)

Can you help with the indicator: stochastic momentum index
How can I get that..?

If it's not available in our library I can help you share the calculations.

As I read at someplace Stochastic Oscillator vs. Stochastic Momentum Index: are different indicators.

https://www.investopedia.com/ask/answers/021315/what-difference-between-stochastic-oscillator-stochastic-momentum-index.asp

Reference Parameters:
![image](https://user-images.githubusercontent.com/76814842/142190791-896904c6-6af9-44e5-ad88-133a67213023.png)

**Comment by DaveSkender** [comment]
I see the original author is William Blau.  Do you have any references to his original description?

**Reply by myalgomate** [reply]
https://www.stockmaniacs.net/stochastic-momentum-index-indicator/

https://patternswizard.com/stochastic-momentum-index/

**Reply by DaveSkender** [reply]
I see one of these mentions that:

> In January 1993, William Blau originated Stochastic Momentum Index publication of “Technical Analysis of Stocks & Commodities” magazine.

I am able to find the original publication in their archive, thanks!  I like to use original sources, rather than recreations by someone else as they tend to _get creative_ with their implementations and occasionally don't implement correctly.

**Reply by DaveSkender** [reply]
I've added this to the backlog, you can track at #626

**Reply by myalgomate** [reply]
Thanks @DaveSkender Will be waiting for the same.

**Reply by myalgomate** [reply]
https://library.tradingtechnologies.com/trade/chrt-ti-stochastic-momentum-index.html

one more reference just in case u need to refer.

**Reply by DaveSkender** [reply]
> https://library.tradingtechnologies.com/trade/chrt-ti-stochastic-momentum-index.html

This is actually a good example of why I don't use third-party reproductions of formulas.  There are significant errors in both what they're describing and the formula itself (they're showing and describing a regular Stochastic Oscillator formula).

The original author's publication, published in the magazine described above by William Blau also uses EMA, not SMA and the formula is similar but different.  With that said, I've implemented the same 4 parameter profile.

The PR (#644) is open for review, if interested, and contains more implementation details.

**Reply by myalgomate** [reply]
Can we add an option to Select MA Type as well..??

![image](https://user-images.githubusercontent.com/76814842/143378023-db472f02-ccae-42ae-a742-202ef90d56f9.png)

**Reply by DaveSkender** [reply]
I see that implementation targets the signal line only, which is probably ok since it’s not messing with the core indicator formula.  Let me think about it a little bit because that would require a different implementation that would operate much slower with a second pass over history. I may want to look at that in a separate effort.

UPDATE: with our addition of [chaining](https://dotnet.stockindicators.dev/guide/#generating-indicator-of-indicators) in v2, you can now generates these and many more overlays as desired.

**Reply by myalgomate** [reply]
Okay, no problem. It was just my thought which I shared with you because most of the platforms provide that feature for selecting MA Type.

**Reply by myalgomate** [reply]
Hi, 
When trying to download the latest library and use GetSMI not able to get the function in the library.
Can you please check if the latest code is available in the library of NuGet package manager

**Reply by DaveSkender** [reply]
Yah, sorry, the doc site sometimes gets ahead of the actual published library.  SMI will be available in the next package, after v1.19.2

**Reply by myalgomate** [reply]
Okay no problem any tentative Date when we will get next version with latest updates.?

**Reply by DaveSkender** [reply]
I’ll probably do it after #652.  Might take a couple weeks or more to finish.

**Reply by DaveSkender** [reply]
This was implemented in version 1.20.0

**Comment by fernaramburu** [comment]
Hi, I found this indicator to be great inside a Trading Platform with all default values but I cannot find the default values on SMI implementation. Can anyone let me know so I can run some test and compare it with what I'm seeing at the trading platform.
Thanks a lot for this question and for the amazing tool you've built!

**Reply by DaveSkender** [reply]
I'd have to dig through my notes and possibly original descriptions from William Blau's original sources to confirm; however, here's what I have set for defaults in our [demo site](https://charts.stockindicators.dev).

![image](https://github.com/DaveSkender/Stock.Indicators/assets/8432125/7b4860a6-5fb1-4d1c-86d4-92b2826c3c8b)

UPDATE: see more recent information on Blau’s “basic configuration” below, he uses `SMI(13,25,2,3-12)`.

**Reply by fernaramburu** [reply]
I guess nobody can ask for a more clear and faster answer. Amazing response time, couldn't ask for more :) Thanks a lot!
I will dig into the code because I still can't understand the differences between this SMI and the SMI from the trading platform.
Not as a critic at all, as I love this library, but this SMI is looking more like a SMO and not so much as a standard SMI. May I be right?  
Thanks again and again! :)

**Reply by DaveSkender** [reply]
Yes, definitely check the code in both comparators.  We source our formulas directly from original authors sources, where possible.  In this case, I see I’d sourced from an article in Stocks & Commodities V. 11:1 (11-18): Stochastic Momentum by William Blau.

**Reply by fernaramburu** [reply]
I just understand it, I was the one looking everything bad. 
Again , great work @DaveSkender . Thanks a lot
I will test deeply test SMI.
BTW, I'm using this amazing library for 1-minute bar to build a live demo and updating the quotes once a minute. Is there any I can ask for the last result calculation only without re-calculating everything every time I call SMI to be calculated? I mean, I feel that it's taking more time because I recalculate N-1 previous SMI values I already did one minute ago...

**Reply by DaveSkender** [reply]
We’re very slowly working on the incremental use case.  Follow #1018 for updates.  The good news, for now, is that recalculating a typical full quote series is still [a fast performance](https://dotnet.stockindicators.dev/performance/), even if it feels a bit icky.  It’s 65us (0.065ms) for SMI for a 500 period quote set.

**Reply by DaveSkender** [reply]
I just re-read William Blau’s article.  His “basic configuration” is:

> SMI (13, 25, 2) means a q = 13-bar stochastic lookback with r = 25-bar first smoothing, s = two-bar second smoothing [and signal] in the range of three to 12 bars.

On our demo site, `SMI(13,25,2,9)` looks like:

<img width="744" alt="image" src="https://github.com/DaveSkender/Stock.Indicators/assets/8432125/7510c228-4e81-4df5-9c41-226a3b3fcb97">
